digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java

```diff
@@ -626,11 +626,117 @@ public void migrate(Handle handle)
+
+            handle.update(
+                    new CreateTableBuilder("queued_task_locks")
+                    .addLongId("id")  // references queued_tasks.id
+                    .addInt("site_id", "")
+                    .addInt("queue_id", "")
+                    .addInt("priority", "not null")
+                    .addInt("retry_count", "not null default 0")
+                    .addLong("lock_expire_time", "")
+                    .addString("lock_agent_id", "")
+                    .build());
+
+            handle.update("insert into queued_task_locks" +
+                    " (id, site_id, queue_id, priority)" +
+                    " select id, site_id, NULL, priority" +
+                    " from queued_tasks");
+
+            // queues
+            handle.update("alter table queues" +
+                    " add column shared_site_id int");
+
+            // resource_types
+            handle.update("drop table resource_types");
+
+            // queued_tasks
+            handle.update("alter table queued_tasks" +
+                    " alter column task_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " alter column queue_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " drop column priority");
+            handle.update("alter table queued_tasks" +
+                    " drop column resource_type_id");
+            handle.update("create unique index queued_tasks_on_site_id_task_id on queued_tasks (site_id, task_id)");
+
+            if (isPostgres()) {
+                handle.update(
+                    "CREATE FUNCTION lock_shared_tasks(target_site_id int, target_site_max_concurrency bigint, limit_count int, lock_expire_seconds int, agent_id text) returns setof bigint as $$\n" +
+                    "BEGIN\n" +
+                    "  IF pg_try_advisory_xact_lock(23300, target_site_id) THEN\n" +
+                    "    RETURN QUERY\n" +
+                    "      with updated as (\n" +
+                    "        update queued_task_locks\n" +
+                    "        set lock_expire_time = cast(extract(epoch from statement_timestamp()) as bigint) + lock_expire_seconds,\n" +
+                    "            lock_agent_id = agent_id\n" +
+                    "        where id = any(\n" +
+                    "          select queued_task_locks.id\n" +
+                    "          from queued_task_locks\n" +
+                    "          where lock_expire_time is null\n" +
+                    "          and site_id = target_site_id\n" +
+                    "          and not exists (\n" +
+                    "            select * from (\n" +
+                    "              select queue_id, count(*) as count\n" +
+                    "              from queued_task_locks\n" +
+                    "              where lock_expire_time is not null\n" +
```

Comment:
I avoided it because it disables index-only scan (on PostgreSQL 9.6). `lock_expire_time IS NOT NULL` is included in a partial index but `lock_expire_time < ...` condition can't be evaluated using `queued_tasks_shared_grouping` or `queued_tasks_ordering` index. Thus it needs to fetch heap pages.

Effects of not having the condition are not big because `expireLocks` is called periodically.

Query plan without `lock_expire_time < ...`:

```
digdag_test=# explain analyze
select queued_task_locks.id
from queued_task_locks
where lock_expire_time is null
and site_id = 1
and not exists (
  select * from (
    select queue_id, count(*) as count
    from queued_task_locks
    where lock_expire_time is not null
      and site_id = 1
    group by queue_id
  ) runnings
  join queues on queues.id = runnings.queue_id
  where runnings.count >= queues.max_concurrency
    and runnings.queue_id = queued_task_locks.queue_id
)
and not exists (
  select count(*)
  from queued_task_locks
  where lock_expire_time is not null
    and site_id = 1
  having count(*) >= 10
)
order by queue_id, priority desc, id
limit 3
;
                                                                                      QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=6.32..25.92 rows=1 width=16) (actual time=0.010..0.010 rows=0 loops=1)
   InitPlan 1 (returns $0)
     ->  Aggregate  (cost=5.90..5.91 rows=1 width=8) (actual time=0.007..0.007 rows=0 loops=1)
           Filter: (count(*) >= 10)
           Rows Removed by Filter: 1
           ->  Index Only Scan using queued_tasks_shared_grouping on queued_task_locks queued_task_locks_2  (cost=0.13..5.90 rows=1 width=0) (actual time=0.004..0.004 rows=0 loops=1)
                 Index Cond: (site_id = 1)
                 Heap Fetches: 0
   ->  Result  (cost=0.41..20.01 rows=1 width=16) (actual time=0.009..0.009 rows=0 loops=1)
         One-Time Filter: (NOT $0)
         ->  Merge Anti Join  (cost=0.41..20.01 rows=1 width=16) (actual time=0.001..0.001 rows=0 loops=1)
               Merge Cond: (queued_task_locks.queue_id = queued_task_locks_1.queue_id)
               ->  Index Only Scan using queued_tasks_ordering on queued_task_locks  (cost=0.12..5.89 rows=1 width=16) (actual time=0.001..0.001 rows=0 loops=1)
                     Index Cond: (site_id = 1)
                     Heap Fetches: 0
               ->  Materialize  (cost=0.28..14.10 rows=1 width=4) (never executed)
                     ->  Nested Loop  (cost=0.28..14.10 rows=1 width=4) (never executed)
                           ->  GroupAggregate  (cost=0.13..5.91 rows=1 width=12) (never executed)
                                 Group Key: queued_task_locks_1.queue_id
                                 ->  Index Only Scan using queued_tasks_shared_grouping on queued_task_locks queued_task_locks_1  (cost=0.13..5.90 rows=1 width=4) (never executed)
                                       Index Cond: (site_id = 1)
                                       Heap Fetches: 0
                           ->  Index Scan using queues_pkey on queues  (cost=0.15..8.17 rows=1 width=8) (never executed)
                                 Index Cond: (id = queued_task_locks_1.queue_id)
                                 Filter: ((count(*)) >= max_concurrency)
 Planning time: 0.318 ms
 Execution time: 0.090 ms
(27 rows)
```


Query plan with `lock_expire_time < ...`:

```
digdag_test=# explain analyze
select queued_task_locks.id
from queued_task_locks
where lock_expire_time is null
and site_id = 1
and not exists (
  select * from (
    select queue_id, count(*) as count
    from queued_task_locks
    where lock_expire_time is not null
      and lock_expire_time < cast(extract(epoch from statement_timestamp()) as bigint)
      and site_id = 1
    group by queue_id
  ) runnings
  join queues on queues.id = runnings.queue_id
  where runnings.count >= queues.max_concurrency
    and runnings.queue_id = queued_task_locks.queue_id
)
and not exists (
  select count(*)
  from queued_task_locks
  where lock_expire_time is not null
    and lock_expire_time < cast(extract(epoch from statement_timestamp()) as bigint)
    and site_id = 1
  having count(*) >= 10
)
order by queue_id, priority desc, id
limit 3
;
                                                                                    QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=6.33..25.94 rows=1 width=16) (actual time=0.011..0.011 rows=0 loops=1)
   InitPlan 1 (returns $0)
     ->  Aggregate  (cost=5.91..5.92 rows=1 width=8) (actual time=0.007..0.007 rows=0 loops=1)
           Filter: (count(*) >= 10)
           Rows Removed by Filter: 1
           ->  Index Scan using queued_tasks_shared_grouping on queued_task_locks queued_task_locks_2  (cost=0.13..5.91 rows=1 width=0) (actual time=0.006..0.006 rows=0 loops=1)
                 Index Cond: (site_id = 1)
                 Filter: (lock_expire_time < (date_part('epoch'::text, statement_timestamp()))::bigint)
   ->  Result  (cost=0.41..20.02 rows=1 width=16) (actual time=0.011..0.011 rows=0 loops=1)
         One-Time Filter: (NOT $0)
         ->  Merge Anti Join  (cost=0.41..20.02 rows=1 width=16) (actual time=0.002..0.002 rows=0 loops=1)
               Merge Cond: (queued_task_locks.queue_id = queued_task_locks_1.queue_id)
               ->  Index Only Scan using queued_tasks_ordering on queued_task_locks  (cost=0.12..5.89 rows=1 width=16) (actual time=0.002..0.002 rows=0 loops=1)
                     Index Cond: (site_id = 1)
                     Heap Fetches: 0
               ->  Materialize  (cost=0.28..14.11 rows=1 width=4) (never executed)
                     ->  Nested Loop  (cost=0.28..14.11 rows=1 width=4) (never executed)
                           ->  GroupAggregate  (cost=0.13..5.92 rows=1 width=12) (never executed)
                                 Group Key: queued_task_locks_1.queue_id
                                 ->  Index Scan using queued_tasks_shared_grouping on queued_task_locks queued_task_locks_1  (cost=0.13..5.91 rows=1 width=4) (never executed)
                                       Index Cond: (site_id = 1)
                                       Filter: (lock_expire_time < (date_part('epoch'::text, statement_timestamp()))::bigint)
                           ->  Index Scan using queues_pkey on queues  (cost=0.15..8.17 rows=1 width=8) (never executed)
                                 Index Cond: (id = queued_task_locks_1.queue_id)
                                 Filter: ((count(*)) >= max_concurrency)
 Planning time: 0.447 ms
 Execution time: 0.063 ms
(27 rows)
```